grace/models/classifier.py

```diff
@@ -60,24 +62,90 @@ def forward(
+        ----------
+        x : torch.Tensor
+            Node features.
+        edge_index : torch.Tensor
+            Edge indices.
+
+        Returns
+        -------
+        node_embeddings : torch.Tensor
+            Learnt features of each node after graph convolutions.
+        edge_embeddings : torch.Tensor
+            Concatenated embeddings of two nodes forming and edge.
+        node_x : torch.Tensor
+            Logit predictions of each node class.
+        edge_x : torch.Tensor
+            Logit predictions of each edge class.
+        """
+        # Ensure the model is in training mode:
+        self.train()
+
+        # Run through a series of graph convolutional layers:
+        # for layer in range(len(self.conv_layer_list)):
+        #     x = self.conv_layer_list[layer](x, edge_index)
+        #     if layer < len(self.conv_layer_list) - 1:
+        #         x = x.relu()
+        #     else:
+        #         node_embeddings = x
+        node_embeddings = x
+
+        # Implement dropout at set probability:
+        node_embeddings = F.dropout(
+            node_embeddings, p=self.dropout, training=self.training
         )
-        node_x = self.node_classifier(embeddings)
 
+        node_x = self.node_classifier(node_embeddings)
+
+        # Get the node embeddings contributing to each edge:
         src, dst = edge_index
-        edge_features = torch.cat(
-            [embeddings[..., src, :], embeddings[..., dst, :]], axis=-1
+        edge_embeddings = torch.cat(
+            [node_embeddings[..., src, :], node_embeddings[..., dst, :]],
+            axis=-1,
         )
 
-        edge_x = self.edge_classifier(edge_features)
+        # Classify the edges through a linear layer:
+        edge_x = self.edge_classifier(edge_embeddings)
+
+        return node_embeddings, edge_embeddings, node_x, edge_x
+
+    def predict(
+        self,
+        x: torch.Tensor,
+        edge_index: torch.Tensor,
+    ) -> tuple[torch.Tensor]:
+        """Perform inference on input data.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            Node features.
+        edge_index : torch.Tensor
+            Edge indices.
+
+        Returns
+        -------
+        node_embeddings : torch.Tensor
+            Learnt features of each node after graph convolutions.
+        edge_embeddings : torch.Tensor
+            Concatenated embeddings of two nodes forming and edge.
+        node_x : torch.Tensor
+            Logit predictions of each node class.
+        edge_x : torch.Tensor
+            Logit predictions of each edge class.
+        """
+
+        # Ensure the model is in evaluation mode:
+        self.eval()
+
+        # Forward pass through the model
+        with torch.no_grad():
+            node_emb, edge_emb, node_x, edge_x = self(x, edge_index)
+            # predicted_classes = torch.argmax(node_x, dim=-1)
 
-        return node_x, edge_x
+        return node_emb, edge_emb, node_x, edge_x
```

Comment:
Just to understand, why are we returning now node_emb, edge_emb? Looks like is not used anywhere and is part of the input right?